Compile fails on FreeBSD due to "undeclared identifier 'UTIME_OMIT'"

Compiling FreeRDP on FreeBSD, with the latest commits, fails with the following errors:

```
/home/giovanni/remmina_devel/FreeRDP/winpr/libwinpr/file/file.c:387:21: 
error: use of undeclared identifier 'UTIME_OMIT'
                 times[0].tv_sec = UTIME_OMIT;
                                   ^
/home/giovanni/remmina_devel/FreeRDP/winpr/libwinpr/file/file.c:388:22: 
error: use of undeclared identifier 'UTIME_OMIT'
                 times[0].tv_nsec = UTIME_OMIT;
                                    ^
/home/giovanni/remmina_devel/FreeRDP/winpr/libwinpr/file/file.c:421:21: 
error: use of undeclared identifier 'UTIME_OMIT'
                 times[1].tv_sec = UTIME_OMIT;
                                   ^
/home/giovanni/remmina_devel/FreeRDP/winpr/libwinpr/file/file.c:422:22: 
error: use of undeclared identifier 'UTIME_OMIT'
                 times[1].tv_nsec = UTIME_OMIT;
                                    ^
/home/giovanni/remmina_devel/FreeRDP/winpr/libwinpr/file/file.c:449:7: 
warning: implicit declaration of function 'futimens' is invalid in C99 
[-Wimplicit-function-declaration]
         rc = futimens(fileno(pFile->fp), times);
              ^
1 warning and 4 errors generated.
```

Is it possible to avoid the UTIME_OMIT identifier (at least on FreeBSD) ?
What about the C99 warning? Can we safely ignore it?

@akallabeth I'm not sure I can find a quick fix as quickly as you would..  Thanks in advance!!!

I've modified it a little bit and I still have a warning because I don't know well how to fix it

[0001-BSD-test-fix_antenore.patch.gz](https://github.com/FreeRDP/FreeRDP/files/157859/0001-BSD-test-fix_antenore.patch.gz)

```sh
/usr/home/tmow/remmina_devel/FreeRDP.patched/winpr/libwinpr/file/file.c:445:34: warning: incompatible pointer types passing 'struct timespec [2]' to parameter of type 'const struct timeval *'
      [-Wincompatible-pointer-types]
        rc = futimes(fileno(pFile->fp), times);
                                        ^~~~~
/usr/include/sys/time.h:483:40: note: passing argument to parameter here
int     futimes(int, const struct timeval *);
                                           ^
```
http://www.unix.com/man-page/freebsd/2/futimes/

futimes expect timeval

More over my patch I think it brakes the rest of your code :-(

At least FreeBSD is compiled at the end.

I'm testing and trying to fix as well.

https://github.com/antenore/FreeRDP/commit/cb990e8c1ead0b0f6970d66f00b9cd2b6ad7a5ea

So far I'm trying to fix this:

```
winpr/libwinpr/file/file.c:386:29: error: no member named 'st_mtimensec' in 'struct stat'
                timevals[0].tv_usec = buf.st_mtimensec / 1000UL;
                                      ~~~ ^
```

--- a/winpr/libwinpr/file/file.c
+++ b/winpr/libwinpr/file/file.c
@@ -352,10 +352,11 @@ static BOOL FileSetFileTime(HANDLE hFile, const FILETIME *lpCreationTime,
 		const FILETIME *lpLastAccessTime, const FILETIME *lpLastWriteTime)
 {
 	int rc;
-#if defined(__APPLE__) || defined(ANDROID)
+#if defined(__APPLE__) || defined(ANDROID) || defined(__FreeBSD__)
 	struct stat buf;
 #endif
-#ifdef ANDROID
+/* OpenBSD, NetBSD and DragonflyBSD support POSIX futimens */
+#if defined(ANDROID) || defined(__FreeBSD__)
 	struct timeval timevals[2];
 #else
 	struct timespec times[2]; /* last access, last modification */
@@ -366,23 +367,30 @@ static BOOL FileSetFileTime(HANDLE hFile, const FILETIME *lpCreationTime,
 	if (!hFile)
 		return FALSE;
 
-#if defined(__APPLE__) || defined(ANDROID)
+#if defined(__APPLE__) || defined(ANDROID) || defined(__FreeBSD__)
 	rc = fstat(fileno(pFile->fp), &buf);
 	if (rc < 0)
 		return FALSE;
 #endif
 	if (!lpLastAccessTime)
 	{
-#ifdef __APPLE__
+#if defined(__APPLE__)
 #if !defined(_POSIX_C_SOURCE) || defined(_DARWIN_C_SOURCE)
 		times[0] = buf.st_atimespec;
 #else
 		times[0].tv_sec = buf.st_atime;
 		times[0].tv_nsec = buf.st_atimensec;
 #endif
-#elif ANDROID
-		timevals[0].tv_sec = buf.st_mtime;
-		timevals[0].tv_usec = buf.st_mtimensec / 1000UL;
+#elif defined(__FreeBSD__)
+		timevals[0].tv_sec = buf.st_atime;
+#ifdef _POSIX_SOURCE
+		TIMESPEC_TO_TIMEVAL(&timevals[0], &buf.st_atim);
+#else
+		TIMESPEC_TO_TIMEVAL(&timevals[0], &buf.st_atimespec);
+#endif
+#elif defined(ANDROID)
+		timevals[0].tv_sec = buf.st_atime;
+		timevals[0].tv_usec = buf.st_atimensec / 1000UL;
 #else
 		times[0].tv_sec = UTIME_OMIT;
 		times[0].tv_nsec = UTIME_OMIT;
@@ -395,7 +403,7 @@ static BOOL FileSetFileTime(HANDLE hFile, const FILETIME *lpCreationTime,
 		tmp -= EPOCH_DIFF;
 		tmp /= 10ULL;
 
-#ifdef ANDROID
+#if defined(ANDROID) || defined(__FreeBSD__)
 		tmp /= 10000ULL;
 
 		timevals[0].tv_sec = tmp / 10000ULL;
@@ -414,7 +422,14 @@ static BOOL FileSetFileTime(HANDLE hFile, const FILETIME *lpCreationTime,
 		times[1].tv_sec = buf.st_mtime;
 		times[1].tv_nsec = buf.st_mtimensec;
 #endif
-#elif ANDROID
+#elif defined(__FreeBSD__)
+		timevals[1].tv_sec = buf.st_mtime;
+#ifdef _POSIX_SOURCE
+		TIMESPEC_TO_TIMEVAL(&timevals[1], &buf.st_mtim);
+#else
+		TIMESPEC_TO_TIMEVAL(&timevals[1], &buf.st_mtimespec);
+#endif
+#elif defined(ANDROID)
 		timevals[1].tv_sec = buf.st_mtime;
 		timevals[1].tv_usec = buf.st_mtimensec / 1000UL;
 #else
@@ -429,7 +444,7 @@ static BOOL FileSetFileTime(HANDLE hFile, const FILETIME *lpCreationTime,
 		tmp -= EPOCH_DIFF;
 		tmp /= 10ULL;
 
-#ifdef ANDROID
+#if defined(ANDROID) || defined(__FreeBSD__)
 		tmp /= 10000ULL;
 
 		timevals[1].tv_sec = tmp / 10000ULL;
@@ -443,7 +458,7 @@ static BOOL FileSetFileTime(HANDLE hFile, const FILETIME *lpCreationTime,
 	// TODO: Creation time can not be handled!
 #ifdef __APPLE__
 	rc = futimes(fileno(pFile->fp), times);
-#elif ANDROID
+#elif defined(ANDROID) || defined(__FreeBSD__)
 	rc = utimes(pFile->lpFileName, timevals);
 #else
 	rc = futimens(fileno(pFile->fp), times);
@@ -617,7 +632,7 @@ BOOL IsFileDevice(LPCTSTR lpDeviceName)
 	return TRUE;
 }
 
-HANDLE_CREATOR _FileHandleCreator = 
+HANDLE_CREATOR _FileHandleCreator =
 {
 	IsFileDevice,
 	FileCreateFileA

